dm/dm/master/scheduler/scheduler_test.go

```diff
@@ -2044,3 +2036,101 @@ func (t *testSchedulerSuite) TestOperateValidatorTask() {
 	t.validatorModeMatch(s, subtaskCfg.Name, subtaskCfg.SourceID, config.ValidationFast) // fail to change mode, remain fast
 	t.validatorStageMatch(s, subtaskCfg.Name, subtaskCfg.SourceID, pb.Stage_Running)     // task running with no error
 }
+
+func (t *testSchedulerSuite) TestUpdateSubTasksAndSourceCfg() {
+	defer t.clearTestInfoOperation()
+
+	var (
+		logger       = log.L()
+		s            = NewScheduler(&logger, config.Security{})
+		sourceID1    = "mysql-replica-1"
+		taskName1    = "task-1"
+		workerName1  = "dm-worker-1"
+		workerAddr1  = "127.0.0.1:8262"
+		subtaskCfg1  config.SubTaskConfig
+		keepAliveTTL = int64(5)
+		ctx          = context.Background()
+	)
+	sourceCfg1, err := config.ParseYamlAndVerify(config.SampleSourceConfig)
+	require.NoError(t.T(), err)
+	sourceCfg1.SourceID = sourceID1
+
+	require.NoError(t.T(), subtaskCfg1.Decode(config.SampleSubtaskConfig, true))
+	subtaskCfg1.SourceID = sourceID1
+	subtaskCfg1.Name = taskName1
+	require.NoError(t.T(), subtaskCfg1.Adjust(true))
+
+	// not started scheduler can't update
+	t.True(terror.ErrSchedulerNotStarted.Equal(s.UpdateSubTasks(ctx, subtaskCfg1)))
+	t.True(terror.ErrSchedulerNotStarted.Equal(s.UpdateSourceCfg(sourceCfg1)))
+
+	// start the scheduler
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	t.NoError(s.Start(ctx, t.etcdTestCli))
+
+	// can't update source when source not added
+	t.True(terror.ErrSchedulerSourceCfgNotExist.Equal(s.UpdateSourceCfg(sourceCfg1)))
+
+	subtaskCfg2 := subtaskCfg1
+	subtaskCfg2.Name = "fake name"
+	// can't update subtask with different task name
+	t.True(terror.ErrSchedulerMultiTask.Equal(s.UpdateSubTasks(ctx, subtaskCfg1, subtaskCfg2)))
+
+	// can't update not added subtask
+	t.True(terror.ErrSchedulerTaskNotExist.Equal(s.UpdateSubTasks(ctx, subtaskCfg1)))
+
+	// start worker,add source and subtask
+	t.NoError(s.AddSourceCfg(sourceCfg1))
+	ctx1, cancel1 := context.WithCancel(ctx)
+	defer cancel1()
+	t.NoError(s.AddWorker(workerName1, workerAddr1))
+	go func() {
+		t.NoError(ha.KeepAlive(ctx1, t.etcdTestCli, workerName1, keepAliveTTL))
+	}()
+	// wait for source1 bound to worker1.
+	t.True(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		bounds := s.BoundSources()
+		return len(bounds) == 1 && bounds[0] == sourceID1
+	}))
+	t.NoError(s.AddSubTasks(false, pb.Stage_Running, subtaskCfg1))
+
+	// can't update subtask not in scheduler
+	subtaskCfg2.Name = subtaskCfg1.Name
+	subtaskCfg2.SourceID = "fake source name"
+	t.True(terror.ErrSchedulerSubTaskNotExist.Equal(s.UpdateSubTasks(ctx, subtaskCfg2)))
```

Comment:
this error is SourceNotExist, not SubTaskNotExist

Reply:
this error message of `SourceNotExist ` is `subtasks with name %s for sources %v not exist`

>> thanks for review 😵 a little dizzy when i fix conflicts this afternoon....